### ISSUE TITLE
HTTP Interceptors don't play well with lazy loading

### ISSUE THREAD

**Issue by taz** [ocombe/ocLazyLoad#24]
Rightio, I ran into an issue when creating an HTTP interceptor to handle 401 unauthorised responses globally (ie: global logout in this case).

I eventually found that angular only evaluates (and reverses) the http interceptors array once; at the conclusion of what it considers to be the last module. Obviously this doesn't really gel well with lazy loading.

The only way I could think of to get around it was to make a change to angular itself which I'm hesitant to PR (I just don't have the spare time at the moment to properly test and submit it... that's if they'd even be interested considering we're kind of doing an end-run around the accepted way of loading modules! :-) )

You can see the changes I had to make in this commit:
https://github.com/BenBlazely/angular.js/commit/9cc372fc2e52468c4623191f1d6c5c80a49e4098

If you can think of a better way of doing it that doesn't require mod'ing angular itself, I'm all ears.

If nothing else I thought I'd put this here in case you or someone else has the same issue and needs a quick fix. (even if it is less than ideal).

B.

**Comment by piotrl** [comment]
There is any known better solution for that as the issue is quite old?

Can we catch/handle the moment when module is not loading properly and the error code? (from what I see, events are only for properly loaded modules)

**Comment by ocombe** [comment]
I think that they are just ignored, aren't they ?

**Comment by piotrl** [comment]
Wrongly loaded modules are ignored - correct, and I want to handle / fix such situation (e.g. user not authorized properly - and show him login form).

As long as there are no any event or returned promise (I'm not sure about the second) - I can't do that.

**Comment by ocombe** [comment]
I can probably do that :)

**Comment by piotrl** [comment]
I've made simple workaround for this problem that is pretty transparent.

Basically I've created wrapper for _$ocLazyLoad_ service that simply checks if module can be loaded BEFORE ocLazyLoad kicks in.

Example of such wrapping function:
```
    function load(modulesNames) {
        return isUserLogged().then(() =>
            $ocLazyLoad.load(modulesNames)
        );
    }
```
Instead of using _$ocLazyLoad_ service, always use new wrapper.

In my case, `isUserLogged()` sends _$http_ request checking if user is authenticated, In case of error it can be easily handled by _$httpInterceptor_.

Note that authentication request will be sent each time we ask to load module (even if it is already loaded). This can be optimized by checking `$ocLazyLoad.isLoaded(modulesNames)` before `isUserLogged()` call - it wasn't my case though.
